[DOCS] Getting started with ES|QL queries guide (#101634)
elastic/elasticsearch: docs/reference/esql/esql-get-started.asciidoc

[[esql-getting-started]]
== Getting started with {esql} queries

++++
<titleabbrev>Getting started</titleabbrev>
++++

> This guide shows you can use {esql} to query and aggregate your data.

Reviewer: ```suggestion
This guide shows how you can use {esql} to query and aggregate your data.
```

[source,esql]
----
FROM sample_data
| WHERE message LIKE "Connected*"
----

[discrete]
[[esql-getting-started-more-commands]]
==== More processing commands

There are many other processing commands, like <<esql-keep>> and <<esql-drop>>
to keep or drop columns, <<esql-enrich>> to enrich a table with data from
indices in {es}, and <<esql-dissect>> and <<esql-grok>> to process data. Refer
to <<esql-processing-commands>> for an overview of all processing commands.

[discrete]
[[esql-getting-started-chaining]]
=== Chain processing commands

You can chain processing commands, separated by a pipe character: `|`. Each
processing command works on the output table of the previous command. The result
of a query is the table produced by the final processing command.

image::images/esql/esql-sort-limit.png[Processing commands can be chained,align="center"]

The following example first sorts the table on `@timestamp`, and next limits the
result set to 3 rows:

[source,esql]
----
FROM sample_data
| SORT @timestamp DESC
| LIMIT 3
----

NOTE: The order of processing commands is important. First limiting the result

> set to 3 rows before sorting those 3 rows would most likely return 3 different

Reviewer: Not sure what you meant here: is that the result of `sort | limit` is different than the one of `limit | sort`? If so the statement " First limiting the result
set to 3 rows before sorting those 3 rows would most likely return 3 different rows" would be better off like " First limiting the result set to 3 rows before sorting those 3 rows would most likely return a result that is different than the one where the sorting comes before the limit".

Author: Yes, that's what I was trying to say here. Thanks! I've changed it into your suggestion.

rows.

[discrete]
[[esql-getting-started-eval]]
=== Compute values

> Use the <<esql-eval>> command to append rows to a table, with calculated values.

Author: Good catch! Thanks.

For example, the following query appends a `duration_ms` column. The values in
the column are computed by dividing `event.duration` by 1,000,000. In other
words: `event.duration` converted from nanoseconds to milliseconds.

[source,esql]
----
FROM sample_data
| EVAL duration_ms = event.duration / 1000000.0
----

`EVAL` supports several <<esql-functions,functions>>. For example, to round a
number to the closest number with the specified number of digits, use the
<<esql-round>> function:

[source,esql]
----
FROM sample_data
| EVAL duration_ms = ROUND(event.duration / 1000000.0, 1)
----

[discrete]
[[esql-getting-started-stats]]
=== Calculate statistics

{esql} can not only be used to query your data, you can also use it to aggregate
your data. Use the <<esql-stats-by>> command to calculate statistics. For
example, the median duration:

[source,esql]
----
FROM sample_data
| STATS median_duration = MEDIAN(event.duration)
----

You can calculate multiple stats with one command:

[source,esql]
----
FROM sample_data
| STATS median_duration = MEDIAN(event.duration), max_duration = MAX(event.duration)
----

> Use `BY` to group calculated stats by one or more other values. For example, to

Reviewer: ```suggestion
Use `BY` to group calculated stats by one or more columns. For example, to
```

calculate the median duration per client IP:

[source,esql]
----
FROM sample_data
| STATS median_duration = MEDIAN(event.duration) BY client.ip
----

[discrete]
[[esql-getting-started-histogram]]
=== Create a histogram

To track statistics over time, {esql} enables you to create histograms using the

> <<esql-auto_bucket>> function. `AUTO_BUCKET` creates human-friendly buckets and

Reviewer: ```suggestion
<<esql-auto_bucket>> function. `AUTO_BUCKET` creates human-friendly bucket sizes and
```

execute a policy that links an IP address to an environment ("Development",
"QA", or "Production"):

[source,console]
----
PUT clientips
{
  "mappings": {
    "properties": {
      "client.ip": {
        "type": "keyword"
      },
      "env": {
        "type": "keyword"
      }
    }
  }
}

PUT clientips/_bulk
{ "index" : {}}
{ "client.ip": "172.21.0.5", "env": "Development" }
{ "index" : {}}
{ "client.ip": "172.21.2.113", "env": "QA" }
{ "index" : {}}
{ "client.ip": "172.21.2.162", "env": "QA" }
{ "index" : {}}
{ "client.ip": "172.21.3.15", "env": "Production" }
{ "index" : {}}
{ "client.ip": "172.21.3.16", "env": "Production" }

PUT /_enrich/policy/clientip_policy
{
  "match": {
    "indices": "clientips",
    "match_field": "client.ip",
    "enrich_fields": ["env"]
  }
}

PUT /_enrich/policy/clientip_policy/_execute
----

////
[source,console]
----
DELETE /_enrich/policy/clientip_policy
----
// TEST[continued]
////

After creating and executing a policy, you can use it with the `ENRICH`
command:

[source,esql]
----
FROM sample_data
| KEEP @timestamp, client.ip, event.duration
| EVAL client.ip = TO_STRING(client.ip)

> | ENRICH clientip_policy

Reviewer 2: I'd use the extended ENRICH syntax here, ie `ENRICH policy ON field`, that IMHO is more clear
